specs/AppNotifications/AppNotificationContentSpec/AppNotificationBuilder-spec.md

```diff
@@ -0,0 +1,1197 @@
+
+Then update the AppNotificationProgressBar using AppNotificationProgressData:
+
+```c#
+data.SequenceNumber(2);
+data.Title(L"Katy Perry"); // Binds to {progressTitle} in xml payload
+data.Value(1.0); // Binds to {progressValue} in xml payload
+data.ValueStringOverride(L"2/2 songs"); // Binds to {progressValueString} in xml payload
+data.Status(L"Downloaded!"); // Binds to {progressStatus} in xml payload
+
+auto result = co_await winrt::Microsoft::Windows::AppNotifications::AppNotificationManager::Default().UpdateAsync(data, L"Tag", L"Group");
+```
+
+Example result: ![Progress Bar Example 2](ProgressBarExample2.png)
+
+# Retrieving Arguments
+
+AppNotificationContent and Button return arguments to the activated application when the user clicks
+on the component. Developers can add multiple key/value pairs and Windows App SDK will serialize
+them into a string.
+
+Below is an example with AppNotificationContent and Button:
+
+```c#
+AppNotificationContent(
+    .AddArgument(L"action", L"appNotificationBodyTouch")
+    .AddArgument(L"user", 938163)
+    .AddButton(Button(L"Send")
+        .AddArgument(L"action", L"buttonTouch")
+        .AddArgument(L"user", 938163)));
+```
+
+When the user clicks on the AppNotification body, the resulting argument string is:
+
+-   "action=appNotificationBodyTouch;user=938163"
+
+When the user clicks on the Button, the result argument string is:
+
+-   "action=buttonTouch;callId=938163"
+
+Notice that ';' is used as the delimiter between each argument.
+
+To retrieve the arguments from the serialized string, developers will use
+AppNotificationReceivedEventArgs::Arguments to get the values from the argument string as a map.
+
+Below is an example usage:
+
+```c#
+using namespace winrt::Microsoft::Windows::AppLifecycle;
+using namespace winrt::Microsoft::Windows::AppNotifications;
+
+// Retrieve the activation args using Windows App SDK AppLifecycle functions
+AppActivationArguments args { AppLifecycle::AppInstance::GetCurrent().GetActivatedEventArgs() };
+ExtendedActivationKind kind { args.Kind() };
+
+// Check the activation kind
+if (kind == ExtendedActivationKind::AppNotification)
+{
+    AppNotificationActivatedEventArgs appActivatedArgs { args.Data().as<AppNotificationActivatedEventArgs>() };
+
+    // Retrieve the serialized arguments from AppNotificationActivatedEventArgs
+    winrt::Windows::Foundation::Collections::IMap<winrt::hstring, winrt::hstring> argumentMap { appActivatedArgs.Argument() };
+}
+```
+
+# Full API Details
+
+```c#
+namespace Microsoft.Windows.AppNotifications.Builder
+{
+    runtimeclass AppNotificationTextProperties
+    {
+        // Contains the set of <text> attributes
+        AppNotificationTextProperties();
+
+        AppNotificationTextProperties SetLanguage(String language);
+        AppNotificationTextProperties UsesCallScenarioAlign();
+        AppNotificationTextProperties SetMaxLines(Int32 maxLines);
+
+        String GetXml();
```

Comment:
I see this follows `XmlDocument.GetXml()`, an alternative is to have this derive from `IStringable` that adds `ToString`, suggesting this just in case you were not aware.

Reply:
In the future we want to use the builder to support Adaptive cards which returns a JSON string instead of an XML string. I imagine in the future it will be :
- String GetXml()
- String GetJson()

But maybe this can be GetPayload(enum Format). For now we just have GetXml and plan for GetJson.

Reply:
If we want to be future-looking, should we just have `AppNotification` take the `Builder` directly? For example...

`AppNotificationManager.Show(new AppNotification(builder));`

Then, developer doesn't have to explicitly call GetXml or GetJson? And we call the correct one based on the version of Windows they're on?

Reply:
> if we want to be future-looking, should we just have AppNotification take the Builder directly? For example...

I like the idea to have Show() directly on the builder instead. That means I don't have to go an discover where I take the notification I built. It's right there on the object. Along the same line, you could also create the build right from the manager:
`AppNotificationManager.Default.CreateNotification().AddText("Hello").Show();`

Reply:
This idea is great but what about those properties that aren't part of the Xml syntax? eg. Group, Priority, Tag etc. If we abstract it out this way, we kinda lose those properties right?

Unless we are saying .AddGroup(), AddTag(), .AddPriority() should be part of the builder syntax too

Reply:
@pmpurifoy @loneursid Please take note of namechange here from *ContentBuilder to *Builder and the additional properties that we need to expose.

Reply:
There's two approaches here, Morten's approach where the builder is created off of the AppNotificationManager, and that's the only way to create the AppNotificationBuilder... (and maybe it's called CreateNotification or BuildNotification)

```csharp
AppNotificationManager.Default.BuildNotification() // Returns a AppNotificationBuilder
    .AddText("My first notification")
    .SetTag("notif1")
    .Show();
```

Or, alternatively we could allow directly creating the AppNotificationBuilder, and it's auto-associated with the default AppNotificationManager (and if we need to support non-default AppNotificationManagers, the Show method could have an overload)

```csharp
new AppNotificationBuilder()
    .AddText("My first notification")
    .SetTag("notif1")
    .Show();
```

I like the simplicity of the second approach (less writing), but one thing I like about Morten's approach is it helps guide people to the AppNotificationManager so they'll know where to discover how to clear and remove notifications too. I like either approach, personally I still favor the second approach since it's less verbose, but I like either one :)

We would need to replicate all the properties on AppNotification. But I've actually implemented this second pattern in the Toolkit about 2 years ago and that's what you see on the public docs today. The user testing I did of that, people liked it (and didn't hear any complaints).

Reply:
Removing all GetXmls and returning an AppNotification object instead through AppNotificationBuilder::BuildNotification